pkg/controller/backupinfrastructure/backup_infrastructure_control.go

```diff
@@ -0,0 +1,372 @@
+			if updateErr := c.updateBackupInfrastructureStatus(op, gardenv1beta1.ShootLastOperationStateProcessing, gardenv1beta1.ShootLastOperationTypeDelete, "Deletion of Backup Infrastructure in progress.", 1, nil); updateErr != nil {
+				backupInfrastructureLogger.Errorf("Could not update the BackupInfrastructure status after deletion start: %+v", updateErr)
+				return true, updateErr
+			}
+
+			if deleteErr := c.deleteBackupInfrastructure(op); deleteErr != nil {
+				c.recorder.Eventf(backupInfrastructure, corev1.EventTypeWarning, gardenv1beta1.EventDeleteError, "[%s] %s", operationID, deleteErr.Description)
+				if updateErr := c.updateBackupInfrastructureStatus(op, gardenv1beta1.ShootLastOperationStateError, gardenv1beta1.ShootLastOperationTypeDelete, deleteErr.Description+" Operation will be retried.", 1, deleteErr); updateErr != nil {
+					backupInfrastructureLogger.Errorf("Could not update the BackupInfrastructure status after deletion error: %+v", updateErr)
+					return true, updateErr
+				}
+				return true, errors.New(deleteErr.Description)
+			}
+			c.recorder.Eventf(backupInfrastructure, corev1.EventTypeNormal, gardenv1beta1.EventDeleted, "[%s] Deleted Backup Infrastructure", operationID)
+			if updateErr := c.updateBackupInfrastructureStatus(op, gardenv1beta1.ShootLastOperationStateSucceeded, gardenv1beta1.ShootLastOperationTypeDelete, "Backup Infrastructure has been successfully deleted.", 100, nil); updateErr != nil {
+				backupInfrastructureLogger.Errorf("Could not update the BackupInfrastructure status after deletion successful: %+v", updateErr)
+				return true, updateErr
+			}
+			if err := c.removeFinalizer(op); err != nil {
+				return true, err
+			}
+			return false, nil
+		}
+	}
+
+	// When a BackupInfrastructure deletion timestamp is not set we need to create/reconcile the backup infrastructure.
+	if updateErr := c.updateBackupInfrastructureStatus(op, gardenv1beta1.ShootLastOperationStateProcessing, operationType, "Reconciliation of Backup Infrastructure state in progress.", 1, nil); updateErr != nil {
+		backupInfrastructureLogger.Errorf("Could not update the BackupInfrastructure status after reconciliation start: %+v", updateErr)
+		return true, updateErr
+	}
+	if reconcileErr := c.reconcileBackupInfrastructure(op); reconcileErr != nil {
+		if updateErr := c.updateBackupInfrastructureStatus(op, gardenv1beta1.ShootLastOperationStateError, operationType, reconcileErr.Description+" Operation will be retried.", 1, reconcileErr); updateErr != nil {
+			backupInfrastructureLogger.Errorf("Could not update the BackupInfrastructure status after reconciliation error: %+v", updateErr)
+			return true, updateErr
+		}
+		return true, errors.New(reconcileErr.Description)
+	}
+	if operationType == gardenv1beta1.ShootLastOperationTypeCreate {
+		c.recorder.Eventf(backupInfrastructure, corev1.EventTypeNormal, gardenv1beta1.EventReconciled, "[%s] Reconciled Backup Infrastructure state", operationID)
```

Comment:
Didn't we say that we only want to send events in case of errors?

Reply:
Successful creation is one time activity and it makes sense to send event in such case. Probably in event description i should change it to "Created" instead of "reconciled". 
What do you say? And i think i removed reconcileErr event as well, i'll add that too.

Reply:
Well, I'm still of the opinion that we do not need events for this resource when no error occurred. How does it help us if we see "Create/Reconcile/Delete successful"? What is the exact benefit? In case of shoots we send events with a specific operation ID so that we can later correlate controller-manager logs against this ID easily. However, in the backup infrastructure controller case there doesn't happen much, either the Terraform execution succeeds or fails.

Reply:
Ok. I agree. There isn't any use case of BackupInfrastructure as of now. Will remove it.